// @deepgram/sdk/dist/main/packages/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrerecordedClient = exports.OnPremClient = exports.ManageClient = exports.LiveClient = exports.ListenClient = exports.AbstractWsClient = exports.AbstractRestfulClient = exports.AbstractClient = void 0;
var AbstractClient_1 = require("./AbstractClient");
Object.defineProperty(exports, "AbstractClient", { enumerable: true, get: function () { return AbstractClient_1.AbstractClient; } });
var AbstractRestfulClient_1 = require("./AbstractRestfulClient");
Object.defineProperty(exports, "AbstractRestfulClient", { enumerable: true, get: function () { return AbstractRestfulClient_1.AbstractRestfulClient; } });
var AbstractWsClient_1 = require("./AbstractWsClient");
Object.defineProperty(exports, "AbstractWsClient", { enumerable: true, get: function () { return AbstractWsClient_1.AbstractWsClient; } });
var ListenClient_1 = require("./ListenClient");
Object.defineProperty(exports, "ListenClient", { enumerable: true, get: function () { return ListenClient_1.ListenClient; } });
var LiveClient_1 = require("./LiveClient");
Object.defineProperty(exports, "LiveClient", { enumerable: true, get: function () { return LiveClient_1.LiveClient; } });
var ManageClient_1 = require("./ManageClient");
Object.defineProperty(exports, "ManageClient", { enumerable: true, get: function () { return ManageClient_1.ManageClient; } });
var OnPremClient_1 = require("./OnPremClient");
Object.defineProperty(exports, "OnPremClient", { enumerable: true, get: function () { return OnPremClient_1.OnPremClient; } });
var PrerecordedClient_1 = require("./PrerecordedClient");
Object.defineProperty(exports, "PrerecordedClient", { enumerable: true, get: function () { return PrerecordedClient_1.PrerecordedClient; } });
